[Modi Trushti Sanjaybhai] iP

tsm1820 opened the pull request:
## Duke 
> 
> "Your mind is for having ideas, not holding them." -- David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke is the friendliest chatbot you'll ever meet. It frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ _SUPER_ _FAST_ to use
* as friendly as any of your human friends 

All you need to do is,

1. download it.
2. double-click it.
3. add your tasks.
4. let it manage your tasks and do the rest of the work for you 😉

And it is **FREE**!

Features:
- [x] Managing todos
- [ ] Managing deadlines
- [ ] Managing events
- [ ] Finding tasks using relevant keywords 

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Ui ob = new Ui();
        ob.execInput();
    }
}
```

Commit:
Add Gradle support

tsm1820 committed:
Add Level 1: Greet, Echo, Exit

tsm1820 committed:
Add Level 2: Add, List

tsm1820 committed:
Add Level 3: Mark as Done

tsm1820 committed:
Add Level 4: ToDos, Events, Deadlines

tsm1820 committed:
Add Level 5: Handle Errors

tsm1820 committed:
Add Level 6: Delete feature, Collections and Enums

tsm1820 committed:
TextUiTesting complete

tsm1820 committed:
Add Level 7: Save

tsm1820 committed:
Add Level 8: Dates and Times

tsm1820 committed:
Merge branch 'branch-Level-7'

tsm1820 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Parser.java
#	src/main/java/TaskList.java

tsm1820 committed:
Add A-Packages

tsm1820 committed:
Add JUnit Tests

tsm1820 committed:
Add A-Jar

tsm1820 committed:
Add Level 9: Find

tsm1820 committed:
Add A-JavaDoc

tsm1820 committed:
Add A-CodingStandard

tsm1820 committed:
Merge branch 'branch-Level-9'

tsm1820 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Parser.java

tsm1820 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java

tsm1820 committed:
Make improvements

tsm1820 committed:
Merge branch 'add-gradle-support'

tsm1820 committed:
Add A-Gradle

tsm1820 committed:
Add A-CheckStyle

tsm1820 committed:
Merge branch 'branch-A-CheckStyle'

huizhuansam commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,34 @@
+        TaskList ob = new TaskList("event concert /at 2018-06-18", Duke.Category.EVENT);
+        assertEquals("[E][ ] concert  (at: Jun 18 2018)",ob.toString());
+    }
+
```
Might just be me nitpicking, but I think it is better to remove blank lines at the end of the file. 
Reason: tidiness.

huizhuansam commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,216 @@
+public class Ui {
+
+    /** Boolean variable to exit the program */
+    private boolean bye = false;
```
Perhaps use more intuitive names for boolean variables such as `isBye`? 
Reason: boolean variables should sound/read like truth statements.

huizhuansam commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,216 @@
+    Path filePath = Paths.get("data/duke.txt");
+
+    /** Stores if the path of the data directory in the project folder exists */
+    boolean existsDirPath = Files.exists(dirPath);
```
Should this be named `dirPathExists` instead? 
Reason: this will make conditional statements read more like normal English, especially in line 63 `if(dirPathExists)` instead of `if(existsDirPath)`.

huizhuansam commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,216 @@
+    boolean existsDirPath = Files.exists(dirPath);
+
+    /** Stores if the path of the data file in the data directory in the project folder exists */
+    boolean existsFilePath = Files.exists(filePath);
```
See comment on line 29.

huizhuansam commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,114 @@
+        boolean flag = true;
+        int i=1;
+        for (int j = 0; j < ob.numberOfTasks(); j++) {
+            if(!ob.getTaskFromList(j).getPreExisting()) {
```
I think it is best to adhere to the style where reserved words such as `if` are followed by a whitespace, i.e. `if (...` instead of `if(...`. I noticed the same issue in other files as well.
Reason: tidiness.

huizhuansam commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,114 @@
+     */
+    public void findTaskFromTaskList(TaskList ob, String inputFind) {
+        String trimmedFind = inputFind.split("\\s", 2)[1];
+        boolean flag = true;
```
Hmm, what does `flag` mean?

huizhuansam commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,114 @@
+     * @param inputFind Keyword to find. 
+     */
+    public void findTaskFromTaskList(TaskList ob, String inputFind) {
+        String trimmedFind = inputFind.split("\\s", 2)[1];
```
Nice use of regex. Perhaps the `trimmedFind` can be more descriptive in what it is, or maybe an explanatory comment can help.

huizhuansam commented on src/main/java/duke/InvalidInputException.java:
```diff
@@ -0,0 +1,15 @@
+ */
+public class InvalidInputException extends DukeException {
+
+    /**
```
Just a comment on the Javadoc (applies to all other files), shouldn't there be an empty line between the description and parameter section?

huizhuansam commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,143 @@
+     * Returns whether the task already exists in the hard disk.
+     * @return boolean value of the isPreExisting field of the task.
+     */
+    public boolean getPreExisting() {
```
Hmm, do you think there is a better name for this boolean method? See `Parser.java` line 33, I wasn't quite sure what the method does in the conditional.

huizhuansam commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,114 @@
+     */
+    public void parseTaskList(TaskList ob) {
+        for (int j = 0; j < ob.numberOfTasks(); j++) {
+            if (!ob.getTaskFromList(j).getPreExisting()) {
```
What does `getPreExisting()` do? I assume that it returns an `Object` (perhaps a `Task`), but seeing that it is used in a conditional, it must have returned a boolean?

huizhuansam commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,216 @@
+            }
+        }
+
+        for(int j=0; j< preExistingTasks.size();j++){
```
Perhaps consider leaving whitespace after the semicolons, i.e. `...; j++`.
Reason: tidiness.

huizhuansam commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,216 @@
+
+        while (!bye) {
+
+            String inp = sc.nextLine();
```
Not quite sure what `inp` means, care to clarify?

huizhuansam reviewed (commented):
Hi Modi, I've left some comments on this PR. Feel free to reply to the comments if you need clarification. 

Best,
Hui Zhuan

kvihashini commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,216 @@
+     * of the program flow based on the input
+     * @throws DukeException
+     */
+    public void execInput() throws DukeException, IOException {
```
Could try to [avoid](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods) longer methods that go beyond 30 LOC, maybe you could split it up into multiple methods (eg. if file exists and if file does not exist) :-)

kvihashini commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,216 @@
+                }
+            } else if (inp.startsWith("done ")) {
+                System.out.println("_____________________________________________________________________");
+                System.out.println("    Nice! I've marked this task as done:");
```
Maybe the [varying level of abstractions](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#slap-hard) (eg. methods and assignment statements) could be reduced

kvihashini commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,53 @@
+     * @param textToAdd the text to be added.
+     * @throws IOException
+     */
+    public void writeToFile(String filePath, String textToAdd) throws IOException {
```
Could use [verbs](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#use-nouns-for-things-and-verbs-for-actions) for methods since it is an action (eg. FileWriter) :-)

kvihashini commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,216 @@
+                System.out.println("    Here are the tasks in your list:");
+                parser.parseTaskList(ob);
+                System.out.println("_____________________________________________________________________");
+            } else if (inp.startsWith("todo")) {
```
Maybe you can have classes for ToDo, DeadLine, Event and make them inherit from a Task class for better [structure](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#structure-code-logically)!

kvihashini reviewed (commented):
LGTM! good job :-) just a few code quality nits to fix!

tsm1820 committed:
Add Level-10

tsm1820 committed:
Merge branch 'branch-Level-10'

tsm1820 committed:
Make improvements to the GUI

tsm1820 committed:
Add A-Assertions

At present, any index can be entered to delete or add a task
to the TaskList.

Let's add assertion checks to ensure that all access to the
ArrayList in the TaskList class is within bounds.

tsm1820 committed:
Add A-CodeQuality

The load method in the Storage class had deep nesting which
was affecting the code quality.

Let's fix the method to make it more readable and structure
it logically. Let's fix some other minor issues affecting code
quality too.

tsm1820 committed:
Merge pull request #1 from tsm1820/branch-A-Assertions

Add A-Assertions

tsm1820 committed:
Merge branch 'master' of github.com:tsm1820/ip

tsm1820 committed:
Merge branch 'master' into branch-A-CodeQuality

tsm1820 committed:
Merge pull request #2 from tsm1820/branch-A-CodeQuality

Add A-CodeQuality

tsm1820 committed:
Merge branch 'master' of github.com:tsm1820/ip

tsm1820 committed:
Add C-DetectDuplicates

Let's add a new feature to detect and manage
duplicate tasks that are added to the list.

tsm1820 committed:
Merge branch 'branch-C-DetectDuplicates'

tsm1820 committed:
Add A-CI

tsm1820 committed:
Fix CheckStyle

tsm1820 committed:
Fix CheckStyleTest for the DukeTest class

tsm1820 committed:
Fix the DukeTest class

tsm1820 committed:
Add A-BetterGui

tsm1820 committed:
Add A-UserGuide

tsm1820 committed:
Update User Guide
